Add rules for Sibbo (CMP)

icodebyamanda opened the pull request:
Add rules to: [www.europafm.com <- CMP sibbo (.ES) <- add](https://app.asana.com/0/1201534009035032/1202683416335233/f)

muodov commented on rules/autoconsent/sibbo.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "name": "sibbo",
+  "prehideSelectors": [".sibbo-layout, .sibbo-panel"],
```
just .sibbo-layout seems to be enough

icodebyamanda replied:
.sibbo-layout only works on tested websites.

muodov commented on rules/autoconsent/sibbo.json:
```diff
@@ -0,0 +1,21 @@
+  "prehideSelectors": [".sibbo-layout, .sibbo-panel"],
+  "detectCmp": [{ "exists": "sibbo-cmp-layout" }],
+  "detectPopup": [{ "visible": ".sibbo-layout" }],
+  "optIn": [{ "click": "#acceptAllMain" }],
```
I see that this button has some data attributes, that might be more stable. How about `.sibbo-layout [data-accept-all]`? Would that work?

icodebyamanda replied:
yes it does work.

muodov commented on rules/autoconsent/sibbo.json:
```diff
@@ -0,0 +1,21 @@
+  "optOut": [
+    { "click": ".sibbo-panel__aside__buttons a[data-nav=\"purposes\"]" },
+    { "click": ".sibbo-panel__main__header__actions a[data-focusable=\"reject-all\"]" },
+    { "wait": "500" },
+    { "click": ".sibbo-panel__main__footer__actions a[data-focusable=\"next\"]" },
```
instead of unconditional wait, how about `.sibbo-panel__main__footer__actions a[data-focusable="next"]:not(.sibbo-cmp-button--disabled)`?

icodebyamanda replied:
As per our MM discussion, unconditional wait was replaced by some `waitFor` selectors for now.
In the future, we may want to switch sibbo into a class-based rules to better handle websites that use a iframe (I'll add a task to track that).

muodov commented on rules/autoconsent/sibbo.json:
```diff
@@ -0,0 +1,21 @@
+  "optIn": [{ "click": "#acceptAllMain" }],
+  "optOut": [
+    { "click": ".sibbo-panel__aside__buttons a[data-nav=\"purposes\"]" },
+    { "click": ".sibbo-panel__main__header__actions a[data-focusable=\"reject-all\"]" },
```
this selector returns multiple elements. Could you make it more specific?

icodebyamanda replied:
First encountered `Reject All ` button was replaced by this selector  `"click": ".sibbo-panel__main__header__actions a[data-focusable=\"reject-all\"]" ` (To be found on L17 & 31)

muodov commented on rules/autoconsent/sibbo.json:
```diff
@@ -0,0 +1,21 @@
+    { "click": ".sibbo-panel__aside__buttons a[data-nav=\"purposes\"]" },
+    { "click": ".sibbo-panel__main__header__actions a[data-focusable=\"reject-all\"]" },
+    { "wait": "500" },
+    { "click": ".sibbo-panel__main__footer__actions a[data-focusable=\"next\"]" },
```
On this site, there's only one page: https://www.parador.es/
I'm afraid we'll need to make this rule a bit more complex to handle different cases

icodebyamanda replied:
Rules rewritten to apply conditional branching and thus to include occurrences of 1-page consent banner vs 2-pages ones (intial rules design).

 Note:
- The "If statement" has a selector that only belongs to 1-page banner and appears from page 0 (banner landing interface)
- If wishing to "manually" test that the condition works: `https://www.estadiodeportivo.com/` is a stable example of 1-page consent website. `https://www.todopapas.com/ `(or many others) follows a 2-page consent pattern.

muodov commented on rules/autoconsent/sibbo.json:
```diff
@@ -0,0 +1,21 @@
+    { "click": ".sibbo-panel__main__footer__actions a[data-focusable=\"next\"]" },
+    { "click": ".sibbo-panel__main div[data-view=\"purposesLegInt\"] a[data-focusable=\"reject-all\"]" },
+    { "wait": "500" },
+    { "click": "#purposesLegIntSaveAndExit" }
```
`[data-save-and-exit]` might be more reliable (watch out, it returns multiple elements, so you'll need to make it more specific)

icodebyamanda replied:
Selector replaced by:
- For 1-page banner L23 -> `"click": "[data-view=\"purposes\"] .sibbo-panel__main__footer__actions [data-save-and-exit]"`
- For 2-page banner L46 -> `"click": ".sibbo-panel__main__footer__actions [data-save-and-exit]"`

muodov commented on rules/autoconsent/sibbo.json:
```diff
@@ -0,0 +1,21 @@
+  ],
+  "test": [
+    {
+      "eval": "!document.cookie.includes('_gpi') && !document.cookie.includes('_gads')"
```
judging purely by the names, these cookies might be set by google, not by the CMP itself. How about checking if `euconsent-v2` is set?

icodebyamanda replied:
Eval replaced by the following (L53): `"eval": "window.localStorage.getItem('euconsent-v2')"`

icodebyamanda commented:
@muodov this one should be good to go.
Main changes include:
- changed selectors for more specific data attributes ones
- unconditional wait replaced by conditional wait
- rule rewrittent to apply conditional branching (inclusive of 1-page banner site vs initial 2-pages one)
- eval is getting the euconsent-v2 in the localStorage

Test wise: self-test, local test and jenkins job now work (parador.es was removed because of its TLS certificate failure)

muodov commented on rules/autoconsent/sibbo.json:
```diff
@@ -0,0 +1,56 @@
+  "name": "sibbo",
+  "prehideSelectors": [".sibbo-layout"],
+  "detectCmp": [{ "exists": "sibbo-cmp-layout" }],
+  "detectPopup": [{ "visible": ".sibbo-layout" }],
```
Again, this rule always returns true, even when popup is closed. I'd suggest `{ "visible": "sibbo-cmp-layout" }`

muodov commented on rules/autoconsent/sibbo.json:
```diff
@@ -0,0 +1,56 @@
+{
+  "name": "sibbo",
+  "prehideSelectors": [".sibbo-layout"],
```
Should we use "sibbo-cmp-layout" everywhere instead of ".sibbo-layout"? It seems more reliable to me

muodov commented on rules/autoconsent/sibbo.json:
```diff
@@ -0,0 +1,56 @@
+        "exists": "[data-view=purposes] .sibbo-panel__main__footer__actions [data-save-and-exit]"
+      },
+      "then": [
+        { 
+          "click": ".sibbo-panel__aside__buttons a[data-nav=\"purposes\"]"
+        },
+        { 
+          "click": ".sibbo-panel__main__header__actions a[data-focusable=\"reject-all\"]" 
+        },
```
these steps, and two last steps are the same in both branches. Can we move them out of the "if" block?
The if condition can probably be based on existensce of `.sibbo-panel__main__footer__actions a[data-focusable=\"next\"]:not(.sibbo-cmp-button--disabled)`, in this case. What do you think?

icodebyamanda replied:
Regarding the following selector `.sibbo-panel__main__footer__actions a[data-focusable=\"next\"]:not(.sibbo-cmp-button--disabled)`, it can't be used as the "if condition" since it appears on both models of banners (on the first settings page).

From what I saw, Sibbo has reused the 2-page banner's HTML to build its 1-page banner model. The former being the main pattern their clients are using.

Therefore, it seems they only added one selector to the 1-page banner model to distinguish between the two models.

muodov commented on lib/config.ts:
```diff
@@ -1 +1 @@
-export const enableLogs = false; // change this to enable debug logs
+export const enableLogs = true; // change this to enable debug logs
```
Please don't forget to rollback this change!

muodov commented on rules/autoconsent/sibbo.json:
```diff
@@ -0,0 +1,56 @@
+  ],
+  "test": [
+    {
+      "eval": "window.localStorage.getItem('euconsent-v2')"
```
eval rules must return a boolean. Add `!!` prefix?

icodebyamanda commented:
@muodov 
Main changes include:
- replaced `.sibbo-layout` by `sibbo-cmp-layout` (i.e. prehideSelectors, detectPopup, optIn)
- Added `!!` to the eval rule
- Added more sites to local tests (while Jenkins is running other jobs)

No change was made to the conditional optOut rule, please find comment here: https://github.com/duckduckgo/autoconsent/pull/24#discussion_r973041642

icodebyamanda commented:
@muodov 
Following our catch up, you'd find the latest version of the conditional rule.
Passed the local and Jenkins tests.

muodov reviewed (approved):
LGTM, nice work!
